docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+cache-friendly code.  In other words, we can load the entire 64-byte
+buffer into a 512-bit wide SIMD register and get data-level
+parallelism on all the columnar values packed into the 64-byte
+buffer. Guaranteed padding can also allow certain compilers to
+generate more optimized code directly (e.g. One can safely use Intel's
+``-qopt-assume-safe-padding``).
+
+Byte Order (`Endianness`_)
+---------------------------
+
+The Arrow format is little endian by default.
+
+Serialized Schema metadata has an endianness field indicating
+endianness of RecordBatches. Typically this is the endianness of the
+system where the RecordBatch was generated. The main use case is
+exchanging RecordBatches between systems with the same Endianness.  At
+first we will return an error when trying to read a Schema with an
+endianness that does not match the underlying system. The reference
+implementation is focused on Little Endian and provides tests for
+it. Eventually we may provide automatic conversion via byte swapping.
+
+Array lengths
+-------------
+
+Array lengths are represented in the Arrow metadata as a 64-bit signed
+integer. An implementation of Arrow is considered valid even if it only
+supports lengths up to the maximum 32-bit signed integer, though. If using
+Arrow in a multi-language environment, we recommend limiting lengths to
+2 :sup:`31` - 1 elements or less. Larger data sets can be represented using
+multiple array chunks.
+
+Null count
+----------
+
+The number of null value slots is a property of the physical array and
+considered part of the data structure. The null count is represented
+in the Arrow metadata as a 64-bit signed integer, as it may be as
+large as the array length.
+
+Validity bitmaps
+----------------
+
+Any type can have null value slots, whether primitive or nested type.
+
+An array with nulls must have a contiguous memory buffer, known as the
+validity (or "null") bitmap, large enough to have at least 1 bit for
+each array slot.
+
+Whether any array slot is valid (non-null) is encoded in the respective bits of
+this bitmap. A 1 (set bit) for index ``j`` indicates that the value is not null,
+while a 0 (bit not set) indicates that it is null. Bitmaps are to be
+initialized to be all unset at allocation time (this includes padding).::
+
+    is_valid[j] -> bitmap[j / 8] & (1 << (j % 8))
+
+We use `least-significant bit (LSB) numbering`_ (also known as
+bit-endianness). This means that within a group of 8 bits, we read
+right-to-left: ::
+
+    values = [0, 1, null, 2, null, 3]
+
+    bitmap
+    j mod 8   7  6  5  4  3  2  1  0
+              0  0  1  0  1  0  1  1
+
+Arrays having a 0 null count may choose to not allocate the null
+bitmap. Implementations may choose to always allocate one anyway as a matter of
+convenience, but this should be noted when memory is being shared.
+
+Nested type arrays have their own null bitmap and null count regardless of
+the null count and null bits of their child arrays.
+
+Primitive
```

Comment:
Add a word after it? (primitive arrays, or primitive types, or primitive layout) 
Just "Primitive" reads a bit strange IMO